# Python Asynchronous Programming Basics
Asynchronously: we don't do things sequentially as we would in synchronous code. We want to run code while some other code is waiting. This refers to the occurrence of events independent of the main program flow.

A co-routine is a wrapped version of a function that allows it to run asynchronously. To create a co-routine, define a function to do what you want

```python
import asyncio
async def main():
    print('shav')
```

By writing ```async``` before the function definition, we create a wrapper around this function. Now when you call this function, it returns a **co-routine object** - this is just like a function and can be executed. 

To execute a co-routine, we ```await``` its execution. To use the await keyword, it needs to be inside of an asynchronous function. This means, this will NOT work:

```python
await main()
```

The reason its not running our code is because we haven't created the **async event loop**. In computer science, the event loop is a programming construct that waits for and dispatches events in a program. In python, we need to define it. Handles all the complicated stuff behind the async syntax.

We must start one in whatever thread we're running this asynchronous code in:

```python
asyncio.run(main())
```
Pass the co-routine into ```asyncio.run()```. This co-routine is the entrypoint into our program. What we did here is create the event loop with ```asyncio.run()``` and added the co-routine in and it ran. Another example with await: 

```python
async def foo(text):
    print(text)
    await asyncio.sleep(1)
```
Needed to await the ```asyncio.sleep(1)``` as it is required to run/execute a co-routine. ```asyncio.sleep(1)``` returns something similar to a co-routine. What if now we had: 

```python
import asyncio

async def main():
    print('shav')
    await foo('text')
    print('finished')

async def foo(text):
    print(text)
    await asyncio.sleep(1)

asyncio.run(main())

```
This will return: ```shav, text, finished```

What could we do so that printing 'finished' is not blocked by the foo() co-routine? For this we can create a **task**:

```python
import asyncio

async def main():
    print('shav')
    task = asyncio.create_task(foo('text'))
    print('finished')

async def foo(text):
    print(text)
    await asyncio.sleep(1)

asyncio.run(main())

```
This will return: ```shav, finished, text```

When this task is created, asyncio starts executing it as soon as it can, and allows other code to run while that task is stalling. While it is not running. It executed the finished statement before the foo(text). When the task was created, the ```main()``` function was still running. 

When the task was created, python cannot run the task until execution of whatever function/whatever it is doing takes a break and stops. With Asyncio, only 1 thing can happen at the exact same time. It is just, when something is not doing something, execution can be paused and given back to another function. We can await tasks to make sure it finishes: 

```python
import asyncio

async def main():
    print('shav')
    task = asyncio.create_task(foo('text'))
    await task
    print('finished')

async def foo(text):
    print(text)
    await asyncio.sleep(1)

asyncio.run(main())

```
This blocks us from moving forward in the program until the task is complete. This will return: ```shav, text, finished``` because we awaited the task. 

Instead, if we added some sleep to the main function, we can see how the task behaves:

```python
import asyncio

async def main():
    print('shav')
    task = asyncio.create_task(foo('text'))
    await asyncio.sleep(2)
    print('finished')

async def foo(text):
    print(text)
    await asyncio.sleep(1)

asyncio.run(main())

```
Here, as soon as we hit ```await asyncio.sleep(2)``` the task will start executing. Because we created the task, we run it as soon as we possibly we can. Since we are awaiting something else inside of the function we are currently running (```main()```), we have a delay. As soon as Asyncio sees the delay, it pauses the execution of the ```main()``` function and allows the execution of the ```task``` that wants to run. This means we don't have the processor waiting around doing nothing. Now we get: ```shav, finished, text```

What if:
```python
import asyncio

async def main():
    print('shav')
    task = asyncio.create_task(foo('text'))
    await asyncio.sleep(0.5)
    print('finished')

async def foo(text):
    print(text)
    await asyncio.sleep(10)

asyncio.run(main())

```
Here:
1.  As soon as we hit ```await asyncio.sleep(0.5)```, the task executes as soon as it can. Because we see the delay with ```await```, we pause the execution of this function, and allow the task that wants to run to run. 
2.  The task is allowed to runs
3.  Asyncio hits another ```await asyncio.sleep(10)``` inside the foo co-routine object. So it allows the executioner to give resource back to the other function. 
4.  finished is printed

We have run things concurrently. We don't have to wait for it to be completely done before we carry on. 

## Futures
```python
import asyncio

async def fetch_data():
    print('start fetching')
    await asyncio.sleep(2)
    print('done fetching')
    return|{'data': 1}

async def print_numbers():
    for i in range(10):
    print(i)
    await asyncio.sleep(6.25)

async def main():
    task1 = asyncio.create_task(fetch_data())
    task2 = asyncio.create_task(print_numbers())

```
Here, when main is ran, task1 will return a future. When you create a task, and the co-routine you defined returns a value, this creates a future. like a promise - it is a placeholder for a value that will exist in the future. 

Before you access the value, you must make sure the co-routine is finished. This is where you would await that task:

```python
...

async def main():
    task1 = asyncio.create_task(fetch_data())
    task2 = asyncio.create_task(print_numbers())

    value = await task1
```
Now, we ensure that task1 is finished and returns that value to value variable. If you didn't await it, you would just print - co-routine. Now task2 wouldn not finish - because it was never awaited. 

## Summary:
+ ```async``` creates a co-routine
+ To execute a co-routine, you need to ```await``` it inside of an  **async event loop** - this is something we need to define
+ ```await``` keyword can only be used inside an ```async``` function - inside an actual co-routine. 
+ to prevent blocking of co-routines, you create tasks and place your co-routine inside of it. This will add it to the event loop
    + When this task is created, asyncio starts executing it as soon as it can, and allows other code to run while that task is stalling.
    + Now we can run code concurrently
+ When you create a task and the co-routine you defined returns a value, this creates a future. Before you access the value, you would ```await``` that task:
